python/foglamp/services/core/snapshot.py

```diff
@@ -0,0 +1,118 @@
+            else:
+                self.check_and_delete_plugins_tar_files(snapshot_plugin_dir)
+
+            self._out_file_path = snapshot_plugin_dir
+            self._interim_file_path = snapshot_plugin_dir
+        except (OSError, Exception) as ex:
+            _LOGGER.error("Error in initializing SnapshotPluginBuilder class: %s ", str(ex))
+            raise RuntimeError(str(ex))
+
+    async def build(self):
+        def reset(tarinfo):
+            tarinfo.uid = tarinfo.gid = 0
+            tarinfo.uname = tarinfo.gname = "root"
+            return tarinfo
+        try:
+            snapshot_id = str(int(time.time()))
+            snapshot_filename = "{}-{}.tar.gz".format(SNAPSHOT_PREFIX, snapshot_id)
+            tar_file_name = "{}/{}".format(self._out_file_path, snapshot_filename)
+            pyz = tarfile.open(tar_file_name, "w:gz")
+            try:
+                pyz.add("{}/python/foglamp/plugins".format(_FOGLAMP_ROOT), recursive=True)
+                # C plugins location is different with "make install" and "make"
+                if _FOGLAMP_ROOT == '/usr/local/foglamp':
+                    pyz.add("{}/plugins".format(_FOGLAMP_ROOT), recursive=True, filter=reset)
+                else:
+                    pyz.add("{}/C/plugins".format(_FOGLAMP_ROOT), recursive=True)
+                    pyz.add("{}/plugins".format(_FOGLAMP_ROOT), recursive=True)
+                    pyz.add("{}/cmake_build/C/plugins".format(_FOGLAMP_ROOT), recursive=True)
+            finally:
+                pyz.close()
+        except Exception as ex:
+            _LOGGER.error("Error in creating Snapshot .tar.gz file: %s ", str(ex))
+            raise RuntimeError(str(ex))
+
+        self.check_and_delete_temp_files(self._interim_file_path)
+        _LOGGER.info("Snapshot %s successfully created.", tar_file_name)
+        return snapshot_id, snapshot_filename
+
+    def check_and_delete_plugins_tar_files(self, snapshot_plugin_dir):
+        files = glob.glob("{}/{}*.tar.gz".format(snapshot_plugin_dir, SNAPSHOT_PREFIX))
+        files.sort(key=os.path.getmtime)
+        if len(files) >= _NO_OF_FILES_TO_RETAIN:
+            for f in files[:-2]:
+                if os.path.isfile(f):
+                    os.remove(os.path.join(snapshot_plugin_dir, f))
+
+    def check_and_delete_temp_files(self, snapshot_plugin_dir):
+        # Delete all non *.tar.gz files
+        for f in os.listdir(snapshot_plugin_dir):
+            if not fnmatch.fnmatch(f, '{}*.tar.gz'.format(SNAPSHOT_PREFIX)):
+                os.remove(os.path.join(snapshot_plugin_dir, f))
+
+    def write_to_tar(self, pyz, temp_file, data):
+        with open(temp_file, 'w') as outfile:
+            json.dump(data, outfile, indent=4)
+        pyz.add(temp_file, arcname=basename(temp_file))
+
+    def extract_files(self, pyz):
+        # Since we are storing full path of the files, we need to specify "/" as the path to restore
+        if _FOGLAMP_ROOT == '/usr/local/foglamp':
```

Comment:
I probably was not very clear, when I said a suid script what I meant was a shell script that was set user id on execution, i.e. chmod 4755 .. That way we can make the script very limited in what is does, i.e. only restore a snapshot from the fixed snapshot directory.

Reply:
Thanks @MarkRiddoch a new script scripts/restore_plugins_snapshot has been added with setuid and setgid bits being set during "sudo make install". This has been successfully tested. Hence new sudoers entry for /bin/tar has been removed.